jumanji/environments/routing/pacman/generator.py

```diff
@@ -0,0 +1,163 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+from typing import Any, List
+
+import chex
+import jax.numpy as jnp
+
+from jumanji.environments.routing.pacman.types import Position, State
+
+
+# flake8: noqa: C901
+def generate_maze_from_ascii(maze: List) -> Any:
+    """Generates a numpy maze from ascii"""
+    ascii_maze = maze
+    numpy_maze = []
+    cookie_spaces = []
+    powerup_spaces = []
+    reachable_spaces = []
+    ghost_spawns = []
+    init_targets = []
+    scatter_targets = []
+    player_coords = None
+
+    for x, row in enumerate(ascii_maze):
+        binary_row = []
+        for y, column in enumerate(row):
+            if column == "G":
+                ghost_spawns.append((y, x))
+            if column == "P":
+                player_coords = (y, x)
+            if column == "X":
+                binary_row.append(0)
+            else:
+                binary_row.append(1)
+                cookie_spaces.append((y, x))
+                reachable_spaces.append((y, x))
+                if column == "O":
+                    powerup_spaces.append((y, x))
+                if column == "T":
+                    init_targets.append((y, x))
+                if column == "S":
+                    scatter_targets.append((y, x))
+
+        numpy_maze.append(binary_row)
+
+    return (
+        numpy_maze,
+        cookie_spaces,
+        powerup_spaces,
+        reachable_spaces,
+        ghost_spawns,
+        player_coords,
+        init_targets,
+        scatter_targets,
+    )
+
+
+class Generator(abc.ABC):
+    def __init__(self, maze: List):
+        """Interface for pacman generator.
+
+        Args:
+            maze: ascii repsesentation of maze to create.
+        """
+        self.x_size = jnp.array(0, jnp.int32)
+        self.y_size = jnp.array(0, jnp.int32)
+        self.pellet_spaces = jnp.array([0, 0])
+
+    @abc.abstractmethod
+    def __call__(self, key: chex.PRNGKey) -> chex.Array:
+        """Generate a problem instance.
+
+        Args:
+            key: random key.
+
+        Returns:
+            state: the generated state.
+        """
+
+
+class AsciiGenerator(Generator):
```

Comment:
I understand this generator is deterministic, i.e. it always returns the same initial state. This is okay, but would it be possible to make it clear in the doc? Either environment docstring, generator docstring, readme, ...

Reply:
Sure, I'll update it in all 3 places so anyone using it doesn't miss it 👍